metrics/src/main/java/org/wildfly/extension/metrics/WildFlyMetric.java

```diff
@@ -0,0 +1,86 @@
+
+/*
+ * JBoss, Home of Professional Open Source.
+ * Copyright 2020, Red Hat, Inc., and individual contributors
+ * as indicated by the @author tags. See the copyright.txt file in the
+ * distribution for a full listing of individual contributors.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package org.wildfly.extension.metrics;
+
+import static org.jboss.as.controller.descriptions.ModelDescriptionConstants.FAILURE_DESCRIPTION;
+import static org.jboss.as.controller.descriptions.ModelDescriptionConstants.NAME;
+import static org.jboss.as.controller.descriptions.ModelDescriptionConstants.OP;
+import static org.jboss.as.controller.descriptions.ModelDescriptionConstants.OP_ADDR;
+import static org.jboss.as.controller.descriptions.ModelDescriptionConstants.READ_ATTRIBUTE_OPERATION;
+import static org.jboss.as.controller.descriptions.ModelDescriptionConstants.RESULT;
+import static org.wildfly.extension.metrics._private.MetricsLogger.LOGGER;
+
+import java.util.OptionalDouble;
+
+import org.jboss.as.controller.LocalModelControllerClient;
+import org.jboss.as.controller.PathAddress;
+import org.jboss.as.controller.descriptions.ModelDescriptionConstants;
+import org.jboss.dmr.ModelNode;
+import org.wildfly.extension.metrics._private.MetricsLogger;
+
+public class WildFlyMetric implements Metric {
+
+    private LocalModelControllerClient modelControllerClient;
+    private final PathAddress address;
+    private final String attributeName;
+
+    public WildFlyMetric(LocalModelControllerClient modelControllerClient, PathAddress address, String attributeName) {
+        this.modelControllerClient = modelControllerClient;
+        this.address = address;
+        this.attributeName = attributeName;
+    }
+
+    @Override
+    public OptionalDouble getValue() {
+        ModelNode result = readAttributeValue(address, attributeName);
+        if (result.isDefined()) {
+            try {
+                return OptionalDouble.of(result.asDouble());
+            } catch (Exception e) {
+                throw MetricsLogger.LOGGER.unableToConvertAttribute(attributeName, address, e);
```

Comment:
This will blow up the entire read.

With Smallrye I believe if a Metric throws an exception it is caught and logged but the loop reading all the metrics continues, so other metrics get reported. A failure here though will abort the PrometheusExporter loop.

Since you are providing an Optional it seems like this could log the failure and fall through.

RBAC failures could be logged at DEBUG/TRACE.

Reply:
done